R/pkg/R/DataFrame.R

```diff
@@ -3406,3 +3406,28 @@ setMethod("randomSplit",
             }
             sapply(sdfs, dataFrame)
           })
+
+#' getNumPartitions
+#'
+#' Return the number of partitions
+#' Note: in order to compute the number of partition the SparkDataFrame has to be converted into a
+#' RDD temporarily internally.
+#'
+#' @param x A SparkDataFrame
+#' @family SparkDataFrame functions
+#' @aliases getNumPartitions,SparkDataFrame-method
+#' @rdname getNumPartitions
+#' @name getNumPartitions
+#' @export
+#' @examples
+#'\dontrun{
+#' sparkR.session()
+#' df <- createDataFrame(cars, numPartitions = 2)
+#' getNumPartitions(df)
+#' }
+#' @note getNumPartitions since 2.1.1
+setMethod("getNumPartitions",
+          signature(x = "SparkDataFrame"),
+          function(x) {
+            getNumPartitionsRDD(toRDD(x))
```

Comment:
Give this is a bit of a hole I think it would be worthwhile to think if there is a reasonable workaround for 2.1.1 release (say JVM wrapper for `.rdd.getNumPartitions`), @shivaram would you agree?

As for the new Scala API, since it has broader implications it might be something to target the 2.2 release? If so that would be better served in a different PR.
I don't mind taking a shot at that - I'm not super familiar with that and from a quick scan it seems to be non-trivial (to handle different RDD subtypes and so on), so a few pointers would be appreciated, @cloud-fan